### PR TITLE
bpo-30987 - Support for ISO-TP protocol in SocketCAN

### PR DESCRIPTION
**Pull request by pylessard**
This Pull Request add the support of CAN ISO-TP address extensions in the socket module by enhancing `getsockaddrarg` and `getsockname`.

**Note to core developpers** : I believe this pull request should be backported back to Python 3.3 where SocketCAN support have been added.

https://bugs.python.org/issue30987

### CODE REVIEW

**Comment by the-knights-who-say-ni** [comment]
Hello, and thanks for your contribution!

I'm a bot set up to make sure that the project can legally accept your contribution by verifying you have signed the [PSF contributor agreement](https://www.python.org/psf/contrib/contrib-form/) (CLA).

Unfortunately we couldn't find an account corresponding to your GitHub username on [bugs.python.org](http://bugs.python.org/) (b.p.o) to verify you have signed the CLA (this might be simply due to a missing "GitHub Name" entry in your b.p.o account settings). This is necessary for legal reasons before we can look at your contribution. Please follow [the steps outlined in the CPython devguide](https://devguide.python.org/pullrequest/#licensing) to rectify this issue.


Thanks again to your contribution and we look forward to looking at it!

**Comment by pylessard** [comment]
It look like the failing test on Travis CI are not related to this Pull Request. The main branch has the same failing tests.

**Comment by pylessard** [comment]
After second thoughts, only the `getsockaddress` interface should be included in Python.
The constant in [can-isotp](https://github.com/hartkopp/can-isotp) should com from a third party module.
Will do the change and reopen the pull request

**Comment by pylessard** [comment]
Pull request is re-open. All dependencies over the third party project have been removed.
`getsockname` interface has been enhanced with ISO-TP protocol support to be consistent with the `bind` method input format.

File: Modules/socketmodule.c [hunk 1]
```diff
@@ -6993,6 +7055,8 @@ PyInit__socket(void)
     PyModule_AddIntMacro(m, CAN_SFF_MASK);
     PyModule_AddIntMacro(m, CAN_EFF_MASK);
     PyModule_AddIntMacro(m, CAN_ERR_MASK);
+
+    PyModule_AddIntMacro(m, CAN_ISOTP);
```
Comment (tiran) [inline Modules/socketmodule.c]: This needs a ``#ifdef CAN_ISOTP``.

File: Modules/socketmodule.c [hunk 2]
```diff
@@ -1913,6 +1927,54 @@ getsockaddrarg(PySocketSockObject *s, PyObject *args,
             Py_DECREF(interfaceName);
             return 1;
         }
+#endif
+
+#if defined(CAN_ISOTP)
```
Comment (tiran) [inline Modules/socketmodule.c]: ``#ifdef CAN_ISOTP``

File: Modules/socketmodule.c [hunk 3]
```diff
@@ -1869,7 +1882,9 @@ getsockaddrarg(PySocketSockObject *s, PyObject *args,
     }
 #endif
 
-#if defined(AF_CAN) && defined(CAN_RAW) && defined(CAN_BCM)
+#if defined(AF_CAN)
```
Comment (tiran) [inline Modules/socketmodule.c]: Please use ifdef for simple checks: ``#ifdef AF_CAN``

File: Lib/test/test_socket.py [hunk 4]
```diff
@@ -1709,6 +1721,49 @@ def testBCM(self):
...
+    def testTooLongInterfaceName(self):
+        # most systems limit IFNAMSIZ to 16, take 1024 to be sure
+        with socket.socket(socket.PF_CAN, socket.SOCK_DGRAM, socket.CAN_ISOTP) as s:
+            self.assertRaisesRegex(OSError, 'interface name too long',
+                                   s.bind, ('x' * 1024,1,2))
+
+    def testBind(self):
+        try:
+            with socket.socket(socket.PF_CAN, socket.SOCK_DGRAM, socket.CAN_ISOTP) as s:
+                addr = (self.interface,0x123,0x456)
```
Comment (tiran) [inline Lib/test/test_socket.py]: Style nit pick: ``addr = self.interface, 0x123, 0x456``

File: Lib/test/test_socket.py [hunk 5]
```diff
@@ -1709,6 +1721,49 @@ def testBCM(self):
...
+                         'socket.CAN_ISOTP required for this test.')
+    def testCreateISOTPSocket(self):
+        with socket.socket(socket.PF_CAN, socket.SOCK_DGRAM, socket.CAN_ISOTP) as s:
+            pass
+
+    def testTooLongInterfaceName(self):
+        # most systems limit IFNAMSIZ to 16, take 1024 to be sure
+        with socket.socket(socket.PF_CAN, socket.SOCK_DGRAM, socket.CAN_ISOTP) as s:
+            self.assertRaisesRegex(OSError, 'interface name too long',
+                                   s.bind, ('x' * 1024,1,2))
```
Comment (tiran) [inline Lib/test/test_socket.py]: Please use

```
with self.assertRaisesRegex(OSError, 'interface name too long'):
    s.bind(('x' * 1024, 1, 2))
```
to make it more readable.

**Review by tiran — CHANGES_REQUESTED** [review]
Thanks for your contribution. Your patch looks mostly fine, good work! I added notes for some style nit picks and a missing ``ifdef`` check.

**Comment by pylessard** [comment]
I didn't expect the Spanish Inquisition!

**Comment by tiran** [comment]
Your PR is still missing a space after comma in ``test_socket.py``, https://github.com/python/cpython/pull/2956/commits/280141c99b81411cd000bb61391a9c39935901c6#diff-30065f853d7944a19b3e17b55db67733 . Let's follow PEP 8 for new code.

**Comment by pylessard** [comment]
@tiran Changes done sir!

**Review by tiran — APPROVED** [review]
perfect! 🥇

**Pull request merged**